src/claim_proofs/ffi_wrapper/src/lib.rs

```diff
@@ -0,0 +1,227 @@
+//! The `claim_proofs_ffi` is the Foreign Function Interface (FFI)
+//! for the `claim_proofs` library. It contains API for generating
+//! claim proofs and verifying them as part of the
+//! Asset Granularity Unique Identity project.
+
+extern crate libc;
+use libc::size_t;
+use std::slice;
+
+use cryptography::claim_proofs::{build_scope_claim_proof_data, compute_cdd_id, compute_scope_id};
+
+pub type ScopeClaimData = cryptography::claim_proofs::ScopeClaimData;
+pub type ScopeClaimProofData = cryptography::claim_proofs::ScopeClaimProofData;
+pub type ProofPublicKey = cryptography::claim_proofs::ProofPublicKey;
+pub type ProofKeyPair = cryptography::claim_proofs::ProofKeyPair;
+pub type CDDClaimData = cryptography::claim_proofs::CDDClaimData;
+pub type Signature = schnorrkel::Signature;
+pub type Scalar = curve25519_dalek::scalar::Scalar;
+pub type RistrettoPoint = curve25519_dalek::ristretto::RistrettoPoint;
+
+// ------------------------------------------------------------------------
+// Data Structures
+// ------------------------------------------------------------------------
+
+#[no_mangle]
+pub extern "C" fn scalar_new(scalar_bits: *const u8, len: size_t) -> *mut Scalar {
+    assert!(!scalar_bits.is_null() && len == 32);
+
+    let mut scalar_slice = [0u8; 32];
+    scalar_slice
+        .copy_from_slice(unsafe { &slice::from_raw_parts(scalar_bits, len as usize)[..32] });
+
+    let scalar = Scalar::from_bits(scalar_slice);
+    Box::into_raw(Box::new(scalar))
+}
+
+#[no_mangle]
+pub extern "C" fn scalar_free(ptr: *mut Scalar) {
+    if ptr.is_null() {
+        return;
+    }
+    unsafe {
+        Box::from_raw(ptr);
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn cdd_claim_data_new(
+    investor_did: *mut Scalar,
+    investor_unique_id: *mut Scalar,
+) -> *mut CDDClaimData {
+    assert!(!investor_did.is_null());
+    assert!(!investor_unique_id.is_null());
+
+    let investor_did: Scalar = unsafe { *investor_did };
+    let investor_unique_id: Scalar = unsafe { *investor_unique_id };
+    Box::into_raw(Box::new(CDDClaimData {
+        investor_did,
+        investor_unique_id,
+    }))
+}
+
+#[no_mangle]
+pub extern "C" fn cdd_claim_data_free(ptr: *mut CDDClaimData) {
+    if ptr.is_null() {
+        return;
+    }
+    unsafe {
+        Box::from_raw(ptr);
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn scope_claim_data_new(
+    scope_did: *mut Scalar,
+    investor_unique_id: *mut Scalar,
+) -> *mut ScopeClaimData {
+    assert!(!scope_did.is_null());
+    assert!(!investor_unique_id.is_null());
+
+    let scope_did: Scalar = unsafe { *scope_did };
+    let investor_unique_id: Scalar = unsafe { *investor_unique_id };
+    Box::into_raw(Box::new(ScopeClaimData {
+        scope_did,
+        investor_unique_id,
+    }))
+}
+
+#[no_mangle]
+pub extern "C" fn scope_claim_data_free(ptr: *mut ScopeClaimData) {
```

Comment:
In general, the points re. `unsafe`, `box_alloc`, and `assert!` apply to everything below here as well to the end of the file.